### ISSUE TITLE
a possible bug?

### ISSUE THREAD

**Issue by daxiongshu** [colinmorris/instacart-basket-prediction#1]
Hi, thank you very much for posting your code. I've spent time reading it and learnt a lot. 
I have this question:

https://github.com/colinmorris/instacart-basket-prediction/blob/4d62ff6c1304639751c305000cf647a85a2d4dac/vectorize.py#L73

This sets the mask based on the maximum cart_index which is not the literal meaning of first_occurences, is it? I thought the mask is to filter out ineligible products that are not ordered yet. if so, shouldn't it be:

```first_occurences = (prev_ordered>0).argmax(axis=1)```

Thank you very much.

**Comment by colinmorris** [comment]
Haha, wow, that's a great catch. You're right, that's a pretty huge bug.

I think `prev_ordered` originally just had 0/1 entries, but later I changed it to store the add-to-cart order and didn't update this line. I really should have written more unit tests/run them more often!

In theory, clipping the loss at some semi-random point in the order history shouldn't necessarily be a big problem. It's sort of like a weaker form of the 'fine-tuning' mode I implemented, where it would only care about the loss on the last order (because that's what the model is tested on in the end). But I found that that fine-tuning mode made training pretty unstable - I assume because of vanishing/exploding gradients. This bug probably did the same thing. Also, the fact that we always start counting at the order in which the focal product was added to the cart latest probably introduces some weird bias.

I did have a vague sense that performance of my models got worse after a big batch of refactoring, and I think this is very likely the culprit!

**Comment by daxiongshu** [comment]
Thank you! I'm wondering what score you could get without this bug.  I'll try to run it and let you know!
